File: src/libs/syscall/src/time/syscall/clock_getres.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Modules
//==================================================================================================

use ::config::kernel::TIMER_FREQ;
use ::sys::{
    error::{
        Error,
        ErrorCode,
    },
    time::NANOSECONDS_PER_SECOND,
};
use ::sysapi::ffi::c_long;
use sysapi::{
    sys_types::clockid_t,
    time::{
        clock_ids::{
            CLOCK_MONOTONIC,
            CLOCK_PROCESS_CPUTIME_ID,
            CLOCK_REALTIME,
            CLOCK_THREAD_CPUTIME_ID,
        },
        timespec,
    },
};

//==================================================================================================
// Standalone Functions
//==================================================================================================

///
/// # Description
///
/// Gets the resolution of the specified clock.
///
/// # Parameters
///
/// - `clock_id`: The clock ID.
/// - `res`: The structure where the resolution is stored.
///
/// # Returns
///
/// Upon successful completion, `clock_getres()` returns empty. Otherwise, it returns an error.
///
pub fn clock_getres(clock_id: clockid_t, res: &mut Option<&mut timespec>) -> Result<(), Error> {
    ::syslog::error!("clock_getres(): clock_id={:?}, res={:?}", clock_id, res);

    match clock_id {
        // Check if the clock ID is valid.
        CLOCK_REALTIME | CLOCK_MONOTONIC => {
            if let Some(res) = res {
                res.tv_sec = 0;
                res.tv_nsec = NANOSECONDS_PER_SECOND as c_long / TIMER_FREQ as c_long;
            }
            Ok(())
        },
        CLOCK_PROCESS_CPUTIME_ID | CLOCK_THREAD_CPUTIME_ID => {
            let reason: &str = "unsupported clock id";
            ::syslog::error!("clock_getres(): {} (clock_id={:?}, res={:?})", reason, clock_id, res);
            Err(Error::new(ErrorCode::OperationNotSupported, "clock_getres() failed"))
        },
        clock_id => {
            let reason: &str = "invalid clock id";
            ::syslog::error!("clock_getres(): {} (clock_id={:?}, res={:?})", reason, clock_id, res);
            Err(Error::new(ErrorCode::InvalidArgument, "clock_getres() failed"))
        },
    }
}

File: src/tests/misc-c/clock_getres.c
/*
 * Copyright(c) The Maintainers of Nanvix.
 * Licensed under the MIT License.
 */

//==================================================================================================
// Imports
//==================================================================================================

#include "common.h"
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

//==================================================================================================
// Standalone Functions
//==================================================================================================

// Tests whether we can get the resolution of a clock with `clock_getres()`.
void test_clock_getres(void)
{
    fprintf(stderr, "testing clock_getres() ... ");

    // Assert size of `timespec` structure.
    STATIC_ASSERT_SIZE(struct timespec, sizeof(time_t) + sizeof(long));

    struct timespec res = {0};

    // Get the resolution of the monotonic clock and check the result.
    memset(&res, 0, sizeof(res));
    assert(clock_getres(CLOCK_MONOTONIC, &res) == 0);
    assert(res.tv_sec >= 0);
    assert(res.tv_nsec >= 0);

    // Get the resolution of the real-time clock and check the result.
    memset(&res, 0, sizeof(res));
    assert(clock_getres(CLOCK_REALTIME, &res) == 0);
    assert(res.tv_sec >= 0);
    assert(res.tv_nsec >= 0);

    // Get the resolution of the process CPU-time clock and check the result.
    memset(&res, 0, sizeof(res));
    assert(clock_getres(CLOCK_PROCESS_CPUTIME_ID, &res) == -1);
    assert(errno == ENOTSUP);
    errno = 0;

    // Get the resolution of the thread CPU-time clock and check the result.
    memset(&res, 0, sizeof(res));
    assert(clock_getres(CLOCK_THREAD_CPUTIME_ID, &res) == -1);
    assert(errno == ENOTSUP);
    errno = 0;

    fprintf(stderr, "passed\n");
}
